Struggling with equally named templates

I am trying to use `rstest-reuse` and it works great. Until I added a second template, with the same name, in a different module of my crate.

It looks like it is creating a macro which gets exported, and thus "pollutes" the root namespace of the crate.

So I tried to define the template in a central location, and re-use it from different sub-modules. Also without success as the `apply` macro panics, as it expects an `Ident` as argument.

Importing the template into the module's namespace using `use` also doesn't work:

~~~
error: macro-expanded `macro_export` macros from the current crate cannot be referred to by absolute paths
  --> src/tests/mqtt/mod.rs:17:5
   |
17 | use crate::mqtt_integration_versions;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: `#[deny(macro_expanded_macro_exports_accessed_by_absolute_paths)]` on by default
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
   = note: for more information, see issue #52234 <https://github.com/rust-lang/rust/issues/52234>
~~~

I am not sure what the right approach here is.

Indeed that is documented when you look more closely. I was focusing this section here: https://github.com/la10736/rstest#use-parametrize-definition-in-more-tests

Regarding dropping the `macro_export` take a look at the following example:

~~~rust
mod foo {
    mod bar {
        use rstest_reuse::{self, *};

        #[template]
        #[rstest(b, case(1))]
        pub fn my_template(#[case] v: i32) {}
    }
    mod baz {
        use rstest_reuse::{self, *};

        #[template]
        #[rstest(b, case(1))]
        pub fn my_template(#[case] v: i32) {}
    }
}
~~~

This will fail with the following error:

~~~
error[E0428]: the name `my_template` is defined multiple times
  --> src/main.rs:16:9
   |
9  |         #[template]
   |         ----------- previous definition of the macro `my_template` here
...
16 |         #[template]
   |         ^^^^^^^^^^^ `my_template` redefined here
   |
   = note: `my_template` must be defined only once in the macro namespace of this module
   = note: this error originates in an attribute macro (in Nightly builds, run with -Z macro-backtrace for more info)
~~~

Which, as you say is expected. But it is only because the macros generated by the `template` macro are exported as well. Which may not be required, depending in your use case.

I made the following change to the template macro (I parsing the argument can probably be done in a better way):

https://github.com/ctron/rstest/blob/e866168f79c52cb33dae6253c0df75af4aa5ffe0/rstest_reuse/src/lib.rs#L204-L232

This now only adds the "export" if "local" is not added. Switching the example above the following now makes it work:

~~~rust
mod foo {
    mod bar {
        use rstest_reuse::{self, *};

        #[template(local)]
        #[rstest(b, case(1))]
        pub fn my_template(#[case] v: i32) {}
    }
    mod baz {
        use rstest_reuse::{self, *};

        #[template(local)]
        #[rstest(b, case(1))]
        pub fn my_template(#[case] v: i32) {}
    }
}
~~~

I personally would prefer adding `(export)` for templates that should be exported. But keeping it compatible makes a lot of sense too.

--- a/rstest_reuse/src/lib.rs
+++ b/rstest_reuse/src/lib.rs
@@ -200,11 +200,26 @@ pub fn merge_attrs(item: TokenStream) -> TokenStream {
 /// attributes. The function signature don't really mater but to make it clear is better that you
 /// use a signature like if you're wrinting a standard `rstest`.
 #[proc_macro_attribute]
-pub fn template(_args: proc_macro::TokenStream, input: proc_macro::TokenStream) -> TokenStream {
+pub fn template(args: proc_macro::TokenStream, input: proc_macro::TokenStream) -> TokenStream {
+    let args: Option<Ident> = parse(args).unwrap();
     let template: ItemFn = parse(input).unwrap();
+
+    let mut tokens = quote! {};
+
+    let local = if let Some(local) = args {
+        local.to_string() == "local"
+    } else {
+        false
+    };
+
+    if !local {
+        tokens.extend(quote! {
+            #[macro_export]
+        });
+    }
+
     let macro_name = template.sig.ident.clone();
-    let tokens = quote! {
-        #[macro_export]
+    tokens.extend(quote! {
         /// Apply #macro_name telmplate to given body
         macro_rules! #macro_name {
             ( $test:item ) => {
@@ -214,7 +229,7 @@ pub fn template(_args: proc_macro::TokenStream, input: proc_macro::TokenStream) 
                         }
                     }
         }
-    };
+    });
     tokens.into()
 }
 

